src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -191,6 +213,60 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
         return new KNNScorer(this, docIdSetIter, scores, boost);
     }
 
+    private BitSet getFilteredDocsBitSet(LeafReaderContext ctx, Weight filterWeight) throws IOException {
+        final Bits liveDocs = ctx.reader().getLiveDocs();
+        final int maxDoc = ctx.reader().maxDoc();
+
+        final Scorer scorer = filterWeight.scorer(ctx);
+        if (scorer == null) {
+            return new FixedBitSet(0);
+        }
+
+        final BitSet acceptDocs = createBitSet(scorer.iterator(), liveDocs, maxDoc);
+        // TODO: Based on this cost shift to exact search, because even in ANN search you have to calculate the
+        // distance for K vectors. This can avoid calls to native layer and save some latency.
+        final int cost = acceptDocs.cardinality();
+        log.info("Number of docs valid for filter is = Cost for filtered k-nn is : {}", cost);
+        return acceptDocs;
+    }
+
+    private BitSet createBitSet(final DocIdSetIterator filteredDocIdsIterator, final Bits liveDocs, int maxDoc) throws IOException {
+        if (liveDocs == null && filteredDocIdsIterator instanceof BitSetIterator) {
+            // If we already have a BitSet and no deletions, reuse the BitSet
+            return ((BitSetIterator) filteredDocIdsIterator).getBitSet();
+        } else {
+            // Create a new BitSet from matching and live docs
+            FilteredDocIdSetIterator filterIterator = new FilteredDocIdSetIterator(filteredDocIdsIterator) {
+                @Override
+                protected boolean match(int doc) {
+                    return liveDocs == null || liveDocs.get(doc);
+                }
+            };
+            return BitSet.of(filterIterator, maxDoc);
+        }
+    }
+
+    private int[] getFilterIdsArray(final LeafReaderContext context) throws IOException {
+        int[] filteredIds = new int[0];
```

Comment:
Can we flip if statement on this line to avoid long nested if?

Reply:
Yes sure,